### PR TITLE
[#138323337] add prefillDeclarationFromFrameworkSlug property to SupplierFramework

### PR DESCRIPTION
**Pull request by risicle**
This sits on top of https://github.com/alphagov/digitalmarketplace-api/pull/525

Story https://www.pivotaltracker.com/story/show/138323337

It was decided that we would reuse/repurpose the `update_supplier_on_framework` endpoint to be a more general `update_supplier_framework`, which now also accepts `reusesAnswersFromFrameworkSlug` being updated. Tests should have better coverage too.

### CODE REVIEW

**Comment by risicle** [comment]
@samuelhwilliams noted that perhaps a better name for this variable might be `prefillAnswersFromFrameworkSlug` and I think people generally agree with that...

**Comment by risicle** [comment]
or even `prefillDeclarationFromFrameworkSlug`... `declarationPrefilledFromFrameworkSlug`?

**Comment by risicle** [comment]
Oh look it's all magically changed name.

File: app/main/views/suppliers.py [hunk 1]
```diff
@@ -418,7 +419,19 @@ def update_supplier_on_framework(supplier_id, framework_slug):
...
+    if "onFramework" in update_json:
+        interest_record.on_framework = update_json["onFramework"]
+    if "prefillDeclarationFromFrameworkSlug" in update_json:
+        if update_json["prefillDeclarationFromFrameworkSlug"] is not None:
+            prefill_declaration_from_framework_id = db.session.query(SupplierFramework.framework_id).filter(
+                SupplierFramework.framework.has(Framework.slug == update_json["prefillDeclarationFromFrameworkSlug"]),
+                SupplierFramework.supplier_id == supplier.supplier_id,
+            ).scalar()
+            if prefill_declaration_from_framework_id is None:
+                abort(400, "Unknown framework slug '{}'".format(update_json["prefillDeclarationFromFrameworkSlug"]))
```
Comment (TheDoubleK) [inline app/main/views/suppliers.py]: It's not quite "Unknown", more "Supplier has not applied for framework slug..."

Comment (risicle) [inline app/main/views/suppliers.py]: Oh that's interesting - I wonder how my change to this message got squeezed out in my rebase. It was supposed to be `Supplier hasn't registered interest in a framework with slug {}`, which is a bit of a cheat because what I'm doing is using the same error for both "unknown framework" and "not registered interest".

File: migrations/versions/830_add_supplierframework_prefill_declaration.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+"""Add SupplierFramework.prefill_declaration_from_framework_id column and constraints
```
Comment (TheDoubleK) [inline migrations/versions/830_add_supplierframework_prefill_declaration.py]: The migration part of this should go in its own separate pull request and get released first on its own.

Comment (risicle) [inline migrations/versions/830_add_supplierframework_prefill_declaration.py]: It is, https://github.com/alphagov/digitalmarketplace-api/pull/525 - it's just that this PR is sitting on top of that because otherwise travis won't pass.

File: tests/conftest.py [hunk 3]
```diff
@@ -222,17 +222,24 @@ def _framework_fixture_inner(request, app, slug, **kwargs):
 
 
 def _supplierframework_fixture_inner(request, app, sf_kwargs=None):
+    """
```
Comment (TheDoubleK) [inline tests/conftest.py]: This is pretty sweet, if slightly mindbending :clap:

Comment (risicle) [inline tests/conftest.py]: Really? I don't like it. I'm just cramming it in there until someone factory boys our api tests...

File: tests/main/views/test_suppliers.py [hunk 4]
```diff
@@ -1511,7 +1516,38 @@ def supplier_framework_interest(self, supplier_framework, update):
                 }),
             content_type='application/json')
 
-    def test_adding_supplier_has_passed(self, supplier_framework):
+    @staticmethod
+    def _refetch_serialized_sf(supplier_framework):
```
Comment (TheDoubleK) [inline tests/main/views/test_suppliers.py]: These also seem really useful.  Nice.

File: tests/main/views/test_suppliers.py [hunk 5]
```diff
@@ -1511,7 +1516,38 @@ def supplier_framework_interest(self, supplier_framework, update):
...
+            ).first(),
+            AuditEvent.type == "supplier_update",
+        ).order_by(AuditEvent.created_at.desc()).first()
+
+    @classmethod
+    def _assert_and_return_audit_event(cls, supplier_framework):
+        # must be performed within an app context
+        audit = cls._latest_supplier_update_audit_event(supplier_framework['supplierId'])
+        assert audit.type == "supplier_update"
+        assert audit.user == "interested@example.com"
```
Comment (TheDoubleK) [inline tests/main/views/test_suppliers.py]: Might it be worth leaving the user check up to the calling test?  I can imagine tests might want to test with different users, maybe?

Comment (risicle) [inline tests/main/views/test_suppliers.py]: I hadn't really intended for this to be used outside `TestSupplierFrameworkUpdates` - if someone finds it useful they can yank it out when they transfer it I guess?

File: tests/main/views/test_suppliers.py [hunk 6]
```diff
@@ -1521,83 +1557,197 @@ def test_adding_supplier_has_passed(self, supplier_framework):
...
+        assert response.status_code == 400
+        error_message = json.loads(response.get_data(as_text=True))['error']
+        assert error_message == "Invalid JSON should not have 'agreementReturned' keys"
+
+        with self.app.app_context():
+            # check nothing has changed on db
+            assert supplier_framework == self._refetch_serialized_sf(supplier_framework)
+            assert self._latest_supplier_update_audit_event(supplier_framework["supplierId"]) is None
+
+    def test_setting_prefill_declaration_from_framework_happy_path(
```
Comment (TheDoubleK) [inline tests/main/views/test_suppliers.py]: `test_setting_and_unsetting...` ?
As this test does both.

File: tests/main/views/test_suppliers.py (hunk 5 above)
Comment (TheDoubleK) [inline tests/main/views/test_suppliers.py]: Fair enough, yep.

File: tests/conftest.py (hunk 3 above)
Comment (TheDoubleK) [inline tests/conftest.py]: Well, it does a job in an interesting way (i.e. I wouldn't have thought of doing it like this).

File: tests/main/views/test_suppliers.py (hunk 6 above)
Comment (risicle) [inline tests/main/views/test_suppliers.py]: yup

**Comment by TheDoubleK** [comment]
Test name isn't a deal breaker - fit it up if you like - and the rest you've addressed already.
:+1:
